posts/inside-rust/2020-02-18-intro-rustc-self-profile.md

```diff
@@ -0,0 +1,253 @@
+| check_impl_item_well_formed                   | 41.36ms   | 0.362           | 77.14ms  | 736        |
++-----------------------------------------------+-----------+-----------------+----------+------------+
+| codegen_fulfill_obligation                    | 40.36ms   | 0.353           | 51.56ms  | 1759       |
++-----------------------------------------------+-----------+-----------------+----------+------------+
+| expand_crate                                  | 37.24ms   | 0.326           | 48.52ms  | 1          |
++-----------------------------------------------+-----------+-----------------+----------+------------+
+| symbol_name                                   | 36.31ms   | 0.317           | 39.06ms  | 5513       |
++-----------------------------------------------+-----------+-----------------+----------+------------+
+| free_global_ctxt                              | 34.34ms   | 0.300           | 34.34ms  | 1          |
++-----------------------------------------------+-----------+-----------------+----------+------------+
+...
+Total cpu time: 11.440758871s
+```
+
+The output is sorted by the self time (time spent in the query or activity but not other queries or activities called by itself).
+As you can see, most of the compilation time is spent in LLVM generating the binary code for the executable.
+
+### `flamegraph`
+
+As you may have guessed, `flamegraph` will produce a [flame graph] of the profiling data.
+To run the tool, we'll pass just the filename without a file extension like we did for `summarize`:
+
+```sh
+$ ../measureme/target/release/flamegraph regex-17088
+```
+
+This will create a file called `rustc.svg` in the working directory:
+
+[![Image of flamegraph output][flame graph img]][flame graph img]
+
+[Click here] to try the interactive svg.
+
+### `crox`
+
+This tool processes self-profiling data into the JSON format that the Chromium profiler understands.
+You can use it to create a graphical timeline showing exactly when various traced events occurred.
+
+In this section, we'll cover a few different modes `crox` can run in such as profiling an entire crate compilation including dependencies and filtering out small events.
+Let's get started with the basics!
+
+#### Basic usage
+
+To run the tool, we'll just pass the filename without a file extension like we've done before:
+
+```sh
+$ ../measureme/target/release/crox regex-17088
+```
+
+This creates a file called `chrome_profiler.json` in the working directory.
+To open it, we'll use the regular Chromium performance tools you might already be familiar with:
+
+1. Open Chrome
+2. Open the Developer Tools console by pressing `Ctrl` + `Shift` + `i` (Windows/Linux) or `Cmd` + `Option` + `i` (macOS)
+3. Click the Performance tab at the top of the console.
+4. Click the "Load profile" button which looks like an arrow pointing up.
+5. Select the `chrome_profiler.json` file we created.
+
+You should now see something similar to this:
```

Comment:
it is also possible to do:
1. Open Chrome
2. Go to chrome://tracing,
3. Click the “Load” button
4. Select the `chrome_profiler.json` file we created.

this will also show the trace file but gives some other possibilities that I have not found for the Developer Tools version. eg. mark one event and gets sum of all events with that name and some more statistics

Reply:
Yes, that's true. I've used that in the past with large trace files which caused the regular tools to choke but it seemed to work fine in this case so I thought it best to keep this as simple as possible. We should document this though probably in the Crox docs.